File: manual/hs/manual/app/Manual/GeneratedNames.hs
{-# LANGUAGE CPP #-}

module Manual.GeneratedNames (examples) where

import Example.Unsafe
import Manual.Tools

{-------------------------------------------------------------------------------
  Examples
-------------------------------------------------------------------------------}

examples :: IO ()
examples = do
    section "Awkward names"

-- There's a quirk with Apple assembler and LLVM IR that do not accept
-- Unicode characters. So make sure to set SUPPORTS_UNICODE environment
-- variable only if you know your system supports it.
#if defined(SUPPORTS_UNICODE)
    -- On supporting platforms, call the functions with Unicode names.
    拜拜
    cϒ
#else
    -- On macOS/LLVM (e.g.), call the safe functions defined in your bindings module.
    -- We assume they are named `gamma` and `byeBye` in Haskell.
    byeBye
    gamma
#endif

    import'

File: manual/c/generated_names.c
#include "generated_names.h"
#include <stdio.h>

/**
 * Awkward names
 */

#if defined(SUPPORTS_UNICODE)
  void 拜拜(void) {
      printf("C function '拜拜' (byebye)\n");
  }

  void ϒ(void) {
      printf("C function 'ϒ' (U+03D2 Greek Upsilon with Hook Symbol)\n");
  }
// Required since Apple's GCC Assembler and LLVM IR does not allow Unicode characters
#else
  void ByeBye() {
      printf("This is the ByeBye function (Unicode-free version).\n");
  }
  void Gamma() {
      printf("This is the Gamma function (Unicode-free version).\n");
  }
#endif

void import(void) {
    printf("C function 'import'\n");
}
